rust library support in builder and conan

partouf opened the pull request:
*cough*

partouf commented:
As this currently is, the following needs to happen to add to CE:
* Libraries need to be added to `rust.amazon.properties` obviously, (also might need to remove some requirements for `path`?)
* Need to add a buildenv that skips all the architecture checks, or implement the rust equivalent
* The decision to download from conan via the buildenv need to be always on, even if not building a binary/executable

partouf commented:
To enable the inclusion of the top 100 packages / the requested packages, I think I'll just write a script that outputs lines for both `libraries.yaml` and for `rust.amazon.properties` to add on

mattgodbolt commented on bin/lib/installation.py:
```diff
@@ -372,6 +373,9 @@ def install(self) -> bool:
         return True
 
     def is_installed(self) -> bool:
+        if self.check_file == None:
```
It's usual in Python do use `is None` (for weird python reasons), or just rely on "truthiness" (if it's ok if `check_file` is `""`) too:

```suggestion
        if self.check_file is None:
```

or

```suggestion
        if not self.check_file:
```

mattgodbolt commented on bin/lib/installation.py:
```diff
@@ -416,13 +420,17 @@ def build(self, buildfor):
-        if self.build_config.build_type == "cmake":
-            return builder.makebuild(buildfor)
-        elif self.build_config.build_type == "make":
+        if self.build_config.build_type in ["cmake", "make"]:
```
nice use of `in` 👍🏻 

(wouldn't it be lovely if python supported `switch` eh? coming soon in python 3.10 is `match`...but one thing at a time)

mattgodbolt commented on bin/lib/rust_library_builder.py:
```diff
@@ -0,0 +1,453 @@
+
+build_timeout = 600
+
+conanserver_url = "https://conan.compiler-explorer.com"
```
I see you updated conan to support `rust` - presumably this is purely to track the build statuses etc?

partouf replied:
this url is indeed just for success/failure + logging + commithash annotation

mattgodbolt commented on bin/lib/rust_library_builder.py:
```diff
@@ -0,0 +1,453 @@
+import contextlib
```
As this is mostly a copy (I think) of existing stuff, I won't go into deep analysis here to save a bunch of "yeah but that's how it is in the other file"

mattgodbolt commented on bin/lib/rust_library_builder.py:
```diff
@@ -0,0 +1,453 @@
+
+    def writeconanfile(self, buildfolder):
+        with (Path(buildfolder) / 'conanfile.py').open(mode='w', encoding='utf-8') as f:
+            f.write('from conans import ConanFile, tools\n')
```
if it's any help at all you can do stuff like:
```py
f.write(f'''from conan import ConanFile, tools
class {self.libname}Conan(ConanFile):
    name = "{self.libname}"
    ...etc...
```

use a single multiline f-string. If you want to keep the indent there are things like https://docs.python.org/3/library/textwrap.html#textwrap.dedent too.

mattgodbolt reviewed (approved):
I'm a bit hazy on the conan side but this seems great! thanks!

partouf commented:
Any thoughts on library name overlapping between c++ / rust? Would this be acceptable, or should we perhaps prefix rust libraries with `rust_` or something?

partouf commented on bin/lib/rust_library_builder.py:
```diff
@@ -0,0 +1,453 @@
+
+        self.logger.debug(f'Buildfolder: {build_folder}')
+
+        source_folder = self.downloadlibrary(build_folder)
```
This is a little intense, we always have to git clone the library to determine the commit hash and compare it with that what was uploaded. Can we somehow circumvent that? Perhaps there is a faster method to get the commit hash from a git repo+branch ?

mattgodbolt replied:
I think some fairly simple parsing of `git ls-remote <gitrepo> <ref>` will do what you want:

```
$ git ls-remote git@github.com:mattgodbolt/performance-tuning.git refs/heads/main
fd7aa6e40e06f09595a56907dffa8f9c5e1cfcb0	refs/heads/main
```

mattgodbolt commented:
> Any thoughts on library name overlapping between c++ / rust? Would this be acceptable, or should we perhaps prefix rust libraries with `rust_` or something? @mattgodbolt

I _think_ it's OK to have the library name overlapping. It would certainly make a lot of things nicer. I think it's just the compilers whose names need to be unique (for legacy reasons mostly I think?)

partouf commented:
> > Any thoughts on library name overlapping between c++ / rust? Would this be acceptable, or should we perhaps prefix rust libraries with `rust_` or something? @mattgodbolt
> 
> I _think_ it's OK to have the library name overlapping. It would certainly make a lot of things nicer. I think it's just the compilers whose names need to be unique (for legacy reasons mostly I think?)

I mean more in context of Conan storage. So say there's a Rust crate called Date with versions 2.0.0 and 3.0.1, there would be conan paths called date/2.0.0 and date/3.0.1 where date/2.0.0 would contain the Rust binaries, and date/3.0.1 would contain both c++ and rust binaries. We can differentiate because we build binaries for different compiler id's, but architecturally it might seem a bit odd.

mattgodbolt commented:
Oh I see...I don't really know about that, I'm not all that familiar with norms for Conan storage...I guess we can try and see? :)

partouf commented:
Whelp.. I just found out that crate versions don't need to be tagged in the github repo. So we can't rely on the git repo.

Downloading the crate source from crates.io seems easy enough, but probably have to use the version number as a commithash to make sure we don't download it 99million times.

partouf commented:
roughly finished what i wanted to do;

* with `ce_install addtoprustcrates` the top 50 crates are requested and added to the yaml file, and a `props` file is generated that can be copy pasted to the `rust.amazon.properties`
* added some caching so the sourcecode isn't downloaded a 100 times
* added a minimum rust compiler version so we skip all the old rust versions to build for

now todo:
* [ ] cleanup old builds in conan & logging
* [ ] build the libraries with the library_builder, which needs some tweaking to its script, etc (https://github.com/compiler-explorer/library-builder/blob/main/build/build.sh#L13)
* [ ] test the rust libraries on staging
